types/react-side-effect/index.d.ts

```diff
@@ -1,23 +1,34 @@
-// Type definitions for react-side-effect v1.0.2
+// Type definitions for react-side-effect 1.1
 // Project: https://github.com/gaearon/react-side-effect
 // Definitions by: Remo H. Jansen <https://github.com/remojansen>
+//                 Martin Charles <https://github.com/0xcaff>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.6
 
 import * as React from "react";
 
-declare function withSideEffect(
-    reducePropsToState: (propsList: any[]) => any,
-    handleStateChangeOnClient: (state: any) => any,
-    mapStateOnServer?: (state: any) => any
-): ClassDecorator;
+interface WithSideEffect {
+    <TProp, TState>(
+        reducePropsToState: (propsList: TProp[]) => TState,
+        handleStateChangeOnClient: (state: TState) => void,
+    ): ClassDecorator<TProp, TState, TState>;
 
-declare class ElementClass extends React.Component<any> {}
-
-interface ClassDecorator {
-    <T extends (typeof ElementClass)>(component:T): T;
+    <TProp, TState, TServerState>(
+        reducePropsToState: (propsList: TProp[]) => TState,
+        handleStateChangeOnClient: (state: TState) => void,
+        mapStateOnServer: (state: TState) => TServerState
+    ): ClassDecorator<TProp, TState, TServerState>;
 }
 
+declare const withSideEffect: WithSideEffect;
+
+type ClassDecorator<TProp, TState, TPeek> = (
+    component: React.ComponentType<TProp>
+) => React.ComponentType<TProp> & {
+    peek: () => TPeek,
+    rewind: () => TState;
```

Comment:
This is wrong, `rewind` always returns `TServerState` and `peek` can returns `TState` or `TServerState`.

I suggest you call the type parameters `TState` and `TServerState` here instead of `TPeek`

Reply:
Good catch, `rewind` always returns `TServerState`. If called on the client, it throws.

https://github.com/gaearon/react-side-effect/blob/f807f02aebc69d0e17428e7bf5886fb83b91312e/src/index.js#L56-L58

`peek` can only return `TState` or `TServerState` based on what context it is called in. I can't return both. `peek` returns `TState` when it is used on the client and `TServerState` when `mapStateOnServer` is provided. https://github.com/gaearon/react-side-effect/blob/f807f02aebc69d0e17428e7bf5886fb83b91312e/src/index.js#L37-L41